File: app/src/main/java/ru/dimagor555/stocks/data/local/stock/LocalStockDatasource.java
package ru.dimagor555.stocks.data.local.stock;

import android.util.Log;
import androidx.paging.Pager;
import androidx.paging.PagingConfig;
import androidx.paging.PagingData;
import androidx.paging.PagingSource;
import androidx.paging.rxjava2.PagingRx;
import io.reactivex.Flowable;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;
import kotlin.jvm.functions.Function0;
import ru.dimagor555.stocks.data.model.stock.Stock;
import ru.dimagor555.stocks.data.model.stock.StockCompanyInfo;
import ru.dimagor555.stocks.data.model.stock.StockPrice;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.util.List;

@Singleton
public class LocalStockDatasource {
    private final StockModelDao dao;
    private final StockModelMapper mapper;

    @Inject
    public LocalStockDatasource(StockModelDao dao, StockModelMapper mapper) {
        this.dao = dao;
        this.mapper = mapper;
    }

    public Flowable<PagingData<Stock>> getAllStocks() {
        return createFlowableStockPagingDataFromPagingSource(dao::getAllStocks);
    }

    public Flowable<PagingData<Stock>> getFavouriteStocks() {
        return createFlowableStockPagingDataFromPagingSource(dao::getFavouriteStocks);
    }

    public Flowable<Boolean> isFavouriteListEmpty() {
        return dao.getFavouriteCount().map(count -> count == 0);
    }

    public Flowable<PagingData<Stock>> getStocksByTickers(List<String> tickers) {
        return createFlowableStockPagingDataFromPagingSource(() ->
                dao.getStocksByTickers(tickers));
    }

    private Flowable<PagingData<Stock>> createFlowableStockPagingDataFromPagingSource(
            Function0<PagingSource<Integer, StockModel>> pagingSource) {
        Pager<Integer, StockModel> pager = new Pager<>(
                new PagingConfig(20, 5, false),
                pagingSource);
        return PagingRx.getFlowable(pager).flatMap(this::pagingDataFromModel);
    }

    private Flowable<PagingData<Stock>> pagingDataFromModel(
            PagingData<StockModel> stockModelPagingData) {
        return Flowable
                .just(PagingRx.map(stockModelPagingData, stockModel ->
                        Single.just(mapper.fromModel(stockModel))));
    }

    public List<String> findTickersByTickerOrCompanyName(String request) {
        return dao.findTickersByTickerOrCompanyName("%" + request + "%");
    }

    public void insertStock(Stock stock) {
        dao.insertStock(mapper.toModel(stock))
                .subscribeOn(Schedulers.io())
                .doOnError(throwable -> Log.e("DB", throwable.getMessage(), throwable))
                .subscribe();
    }

    public void updateStock(Stock stock) {
        dao.updateStock(mapper.toModel(stock))
                .subscribeOn(Schedulers.io())
                .doOnError(throwable -> Log.e("DB", throwable.getMessage(), throwable))
                .subscribe();
    }

    public void updateStockCompanyInfo(String ticker, StockCompanyInfo newCompanyInfo) {
        Stock updatableStock = getStockByTicker(ticker);
        updatableStock.getCompanyInfo()
                .setCompanyInfo(
                        newCompanyInfo.getCompanyName(),
                        newCompanyInfo.getCompanySiteUrl(),
                        newCompanyInfo.getLogoUrl());
        updateStock(updatableStock);
    }

    public void updateStockPriceInfo(String ticker, StockPrice newPrice) {
        Stock updatableStock = getStockByTicker(ticker);
        updatableStock.getPrice()
                .setPrice(
                        newPrice.getCurrPriceInCents(),
                        newPrice.getPreviousClosePriceInCents(),
                        newPrice.getPriceTime());
        updateStock(updatableStock);
    }

    public Stock getStockByTicker(String ticker) {
        return mapper.fromModel(dao.getStockByTicker(ticker));
    }
}

File: app/src/main/java/ru/dimagor555/stocks/data/local/stock/LocalStockDatasource.kt
package ru.dimagor555.stocks.data.local.stock

import androidx.paging.Pager
import androidx.paging.PagingConfig
import androidx.paging.PagingData
import androidx.paging.PagingSource
import androidx.paging.rxjava2.flowable
import androidx.paging.rxjava2.mapAsync
import io.reactivex.Flowable
import io.reactivex.Single
import ru.dimagor555.stocks.data.local.DbUtils
import ru.dimagor555.stocks.data.local.StocksDatabase
import ru.dimagor555.stocks.data.local.stock.dao.StockBaseModelDao
import ru.dimagor555.stocks.data.local.stock.dao.StockModelDao
import ru.dimagor555.stocks.data.local.stock.dao.StockPriceModelDao
import ru.dimagor555.stocks.data.local.stock.entity.StockModel
import ru.dimagor555.stocks.data.local.stock.mapper.StockModelMapper
import ru.dimagor555.stocks.data.model.stock.entity.Stock
import ru.dimagor555.stocks.data.model.stock.entity.StockCompanyInfo
import ru.dimagor555.stocks.data.model.stock.entity.StockPrice
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class LocalStockDatasource @Inject constructor(
    private val database: StocksDatabase,
    private val stockDao: StockModelDao,
    private val stockBaseDao: StockBaseModelDao,
    private val stockPriceDao: StockPriceModelDao,
    private val mapper: StockModelMapper
) {
    val allStocks: Flowable<PagingData<Stock>>
        get() = createFlowableStockPagingDataFromPagingSource { stockDao.allStocks }
    val favouriteStocks: Flowable<PagingData<Stock>>
        get() = createFlowableStockPagingDataFromPagingSource { stockDao.favouriteStocks }
    val isFavouriteListEmpty: Flowable<Boolean>
        get() = stockDao.favouriteCount.map { count: Int -> count == 0 }

    fun getStocksByTickers(tickers: List<String>): Flowable<PagingData<Stock>> {
        return createFlowableStockPagingDataFromPagingSource { stockDao.getStocksByTickers(tickers) }
    }

    fun getStockByTicker(ticker: String): Flowable<Stock?> {
        val stockModelFlowable = stockDao.getStockFlowableByTicker(ticker = ticker)
        return stockModelFlowable.map {
            return@map mapper.fromStockModel(it)
        }
    }

    private fun createFlowableStockPagingDataFromPagingSource(
        pagingSource: Function0<PagingSource<Int, StockModel>>
    ): Flowable<PagingData<Stock>> {
        val pager = Pager(
            PagingConfig(20, 5, false),
            null,
            pagingSource
        )
        return pager.flowable.flatMap { pagingDataFromModel(it) }
    }

    private fun pagingDataFromModel(
        stockModelPagingData: PagingData<StockModel>
    ): Flowable<PagingData<Stock>> {
        return Flowable
            .just(stockModelPagingData.mapAsync {
                Single.just(mapper.fromStockModel(it))
            })
    }

    fun findTickersByTickerOrCompanyName(request: String): List<String> {
        return stockDao.findTickersByTickerOrCompanyName("%$request%")
    }

    fun insertStock(stock: Stock) {
        DbUtils.runDbQuery {
            if (stockBaseDao.hasStockWithTicker(stock.ticker))
                return@runDbQuery

            database.runInTransaction {
                val priceModelToInsert = mapper.toStockPriceModel(stock)
                val priceId = stockPriceDao
                    .insertStockPriceModel(priceModelToInsert).toInt()
                val stockBaseModel = mapper.toStockBaseModel(stock, priceId)
                stockBaseDao.insertStockBaseModel(stockBaseModel)
            }
        }
    }

    fun updateStockFavourite(ticker: String, favourite: Boolean) {
        DbUtils.runDbQuery {
            stockBaseDao.getStockByTicker(ticker)?.let {
                val baseModelToUpdate = it.copy(favourite = favourite)
                stockBaseDao.updateStockBaseModel(baseModelToUpdate)
            }
        }
    }

    fun updateStockCompanyInfo(ticker: String, newCompanyInfo: StockCompanyInfo) {
        DbUtils.runDbQuery {
            stockBaseDao.getStockByTicker(ticker)?.let {
                val baseModelToUpdate =
                    with(newCompanyInfo) {
                        it.copy(
                            companyName = companyName,
                            companySiteUrl = companySiteUrl,
                            logoUrl = logoUrl
                        )
                    }
                stockBaseDao.updateStockBaseModel(baseModelToUpdate)
            }
        }
    }

    fun updateStockPriceInfo(ticker: String, newPrice: StockPrice) {
        DbUtils.runDbQuery {
            stockBaseDao.getStockByTicker(ticker)?.let {
                val oldPrice = stockPriceDao.getStockPriceById(it.priceId)
                val priceModelToUpdate =
                    with(newPrice) {
                        oldPrice.copy(
                            currPriceInCents = currPriceInCents,
                            previousClosePriceInCents = previousClosePriceInCents,
                            priceTime = priceTime
                        )
                    }
                stockPriceDao.updateStockPriceModel(priceModelToUpdate)
            }
        }
    }
}
